rfc/rfc-78/rfc-78.md

```diff
@@ -0,0 +1,339 @@
+- If partial updates are enabled, after downgrade we might fallback to key based merges, giving up on the performance optimization.
+- Any functional indexes built in 1.x might be deleted during downgrade. 
+- Any secondary indexes built in 1.x might be deleted during downgrade. 
+- Completion time based log file ordering within a file slice may not be honored.
+
+### Users will have to use hudi-cli to do the downgrade. Here are the steps we might need to do during downgrade
+- Similar to our upgrade, lets trigger full compaction to latest file slice so that all latest file slices only has base files. We do not want to have any file slices w/ a mix of 1.x log files and 0.x log files. 
+We could afford to take some downtime during downgrade, but upgrade path should be kept to as minimum as possible. So, during upgrade, we are not enforcing this.   
+- Rewrite LSM based timeline to archived timeline. we have to deduce writer properties and introduce boundaries based on that (i.e. until which commit to archive).
+- We have two options wrt handling active timeline instants(for eg, pending clustering instants):
+   A: No changes to active instants. In order to support 1.x tables w/ 0.16.0 reader, already these changes might have been ported to 0.16.0 reader. But we might have to ensure downgrade from 0.16.0 to any older version of hudi does take care of rewriting the active timeline instants.
+   B: While downgrading from 1.x to 0.16.0, let’s rewrite the active timeline instants too. So that after downgrade, we can't differentiate whether table was natively created in 0.16.0 or was it upgraded to 1.x and then later downgraded. The con in Option A is not an issue here since we take care of active timeline instants during 1.x to 0.16.0 downgrade only. So, downgrade from 0.16.0 to any lower hudi version does not need to do any special handling if table was native to 0.16.0 or was it downgraded from 1.x to 0.16.0
+   Our proposal is to go with Option B to ensure 0.16.0 table will be intact and not have any residues from 1.x writes. While the rewrite of active instants are happening, we could see some read failures. But 
+   we do not want to have a 0.16.0 (post upgrade) table to deal w/ 1.x timeline intricacies. 
+- If there are new MDT partitions (functional index, sec index), nuke them. Update table properties.
+- Check latest snapshot files for log files w/ new features enabled. For eg, deletion vector, partial update. If any log files are found in latest snapshot, we have to trigger compaction for those file groups. This calls for a custom compaction strategy as well which compacts only for certain file groups.
+   - In order to reduce downtime to downgrade, we are inclining to do this only for snapshot reads. If there are older file slices and users are issuing time travel or incremental reads, and if they have any of the new features enabled (deletion vector, partial update), readers could break after downgrade.
```

Comment:
An optimization we thought that could help us in the downgrade flow or post downgrade flow: 
say we introduce a table property called "table.version.lineage", whose value is a comma separated list of all diff table versions this table has come through. 

for eg, "7,8". this means that, table was created w/ version 7 and then upgraded to 8. 

"7,8,7" means, table was with 7 and then upgraded to 8 and then downgraded to 7 again. 

so, by interception this config value we can deterministically say if table has been upgraded to 1.x before or not. If yes, we could completely avoid downgrading a 0.16.0 table further down (say 0.15.0.). 

If not, we should be allowing downgrade from 0.16.0 to 0.15.0. Reason is: say someone upgraded from 0.15.0 to 0.16.0 and due to some issues, they want to downgrade to 0.15.0. We should definitely allow them to downgrade. 

only sequence we want to add constraints is: after upgrading to 1.0.x, and downgraded to 0.16.0. After sometime, if they wish to downgrade to 0.15.0, we should avoid the downgrade.